### ISSUE TITLE
wrong code at -O3 on x86_64-linux-gnu

### ISSUE THREAD

**Issue by zhendongsu** [llvm/llvm-project#88103]
It appears to be a recent regression as it does not reproduce with 18.1.0.

Compiler Explorer: https://godbolt.org/z/E39arE65W

```
[587] % clangtk -v
clang version 19.0.0git (https://github.com/llvm/llvm-project.git 51089e360e37962c7841fe0a494ba9fb5368bab2)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /local/suz-local/software/local/clang-trunk/bin
Build config: +assertions
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/10
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/11
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/9
Selected GCC installation: /usr/lib/gcc/x86_64-linux-gnu/11
Candidate multilib: .;@m64
Selected multilib: .;@m64
[588] % 
[588] % clangtk -O2 small.c; ./a.out
[589] % clangtk -O3 small.c
[590] % ./a.out
Aborted
[591] % cat small.c
volatile int a, c;
char b, f, j;
int d, g, h, i;
unsigned char e;
int main() {
  f = i = 0;
  for (; i < 4; i++)
    for (j = 23; j > -18; j--)
      a;
  for (h = 2; h; h--)
    for (b = 0; b < 3; b++) {
      f ^= c;
      d += f >= g;
      g = --e;
    }
  if (d != 1)
    __builtin_abort();
  return 0;
}
```

**Comment by antoniofrighetto** [comment]
Partially reduced to:
```llvm
@i = dso_local local_unnamed_addr global i32 0, align 4
@f = dso_local local_unnamed_addr global i8 0, align 1
@j = dso_local local_unnamed_addr global i8 0, align 1
@c = dso_local global i32 0, align 4
@g = dso_local local_unnamed_addr global i32 0, align 4
@d = dso_local local_unnamed_addr global i32 0, align 4
@e = dso_local local_unnamed_addr global i8 0, align 1

define noundef i32 @main() {
  store i8 0, ptr @f, align 1
  store i8 -18, ptr @j, align 1
  store i32 4, ptr @i, align 4
  %g.promoted46 = load i32, ptr @g, align 4
  %d.promoted48 = load i32, ptr @d, align 4
  %e.promoted50 = load i8, ptr @e, align 1
  %.0 = add i8 %e.promoted50, -3
  %1 = load volatile i32, ptr @c, align 4
  %2 = trunc i32 %1 to i8
  %conv16 = sext i8 %2 to i32
  %cmp17 = icmp sle i32 %g.promoted46, %conv16
  %conv18 = zext i1 %cmp17 to i32
  %add = add nsw i32 %d.promoted48, %conv18
  %dec19 = add i8 %e.promoted50, -1
  %conv20 = zext i8 %dec19 to i32
  %3 = load volatile i32, ptr @c, align 4
  %4 = trunc i32 %3 to i8
  %conv15.1 = xor i8 %2, %4
  %conv16.1 = sext i8 %conv15.1 to i32
  %cmp17.1 = icmp sle i32 %conv20, %conv16.1
  %conv18.1 = zext i1 %cmp17.1 to i32
  %add.1 = add nsw i32 %add, %conv18.1
  %dec19.1 = add i8 %e.promoted50, -2
  %conv20.1 = zext i8 %dec19.1 to i32
  %5 = load volatile i32, ptr @c, align 4
  %6 = trunc i32 %5 to i8
  %conv15.2 = xor i8 %conv15.1, %6
  %conv16.2 = sext i8 %conv15.2 to i32
  %cmp17.2 = icmp sle i32 %conv20.1, %conv16.2
  %conv18.2 = zext i1 %cmp17.2 to i32
  %add.2 = add nsw i32 %add.1, %conv18.2
  %7 = zext i8 %.0 to i32
  %8 = load volatile i32, ptr @c, align 4
  %9 = trunc i32 %8 to i8
  %conv15.157 = xor i8 %conv15.2, %9
  %conv16.158 = sext i8 %conv15.157 to i32
  %cmp17.159 = icmp sle i32 %7, %conv16.158
  %conv18.160 = zext i1 %cmp17.159 to i32
  %add.161 = add nsw i32 %add.2, %conv18.160
  %dec19.162 = add i8 %e.promoted50, -4
  %conv20.163 = zext i8 %dec19.162 to i32
  %10 = load volatile i32, ptr @c, align 4
  %11 = trunc i32 %10 to i8
  %conv15.1.1 = xor i8 %conv15.157, %11
  %conv16.1.1 = sext i8 %conv15.1.1 to i32
  %cmp17.1.1 = icmp sle i32 %conv20.163, %conv16.1.1
  %conv18.1.1 = zext i1 %cmp17.1.1 to i32
  %add.1.1 = add nsw i32 %add.161, %conv18.1.1
  %dec19.1.1 = add i8 %e.promoted50, -5
  %conv20.1.1 = zext i8 %dec19.1.1 to i32
  %12 = load volatile i32, ptr @c, align 4
  %13 = trunc i32 %12 to i8
  %conv15.2.1 = xor i8 %conv15.1.1, %13
  %conv16.2.1 = sext i8 %conv15.2.1 to i32
  %cmp17.2.1 = icmp sle i32 %conv20.1.1, %conv16.2.1
  %conv18.2.1 = zext i1 %cmp17.2.1 to i32
  %add.2.1 = add nsw i32 %add.1.1, %conv18.2.1
  %cmp27.not = icmp eq i32 %add.2.1, 1
  br i1 %cmp27.not, label %if.end, label %if.then

if.then:                                          ; preds = %entry
  tail call void @abort()
  unreachable

if.end:                                           ; preds = %entry
  ret i32 0
}

declare void @abort()
```

**Comment by antoniofrighetto** [comment]
Maybe cc/ @alexey-bataev

**Comment by alexey-bataev** [comment]
Thanks for the report, will investigate it ASAP

**Comment by alexey-bataev** [comment]
Fixed in 910d2de357de8a490cac3ecbd27196356fe1f2a3
